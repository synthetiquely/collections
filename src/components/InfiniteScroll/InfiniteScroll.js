import React, { Component } from 'react';
import { observer } from 'mobx-react';
import Spinner from '../styled/Spinner';
import CenteredContainer from '../styled/CenteredContainer';
import { SCROLL_THESHOLD } from '../../constants';

@observer
class InfiniteScoll extends Component {
  constructor(props) {
    super(props);
    this.container = null;
    this.onScroll = this.onScroll.bind(this);
    this.setRef = this.setRef.bind(this);
    this.loadMore = this.loadMore.bind(this);
  }

  componentDidMount() {
    window.addEventListener('scroll', this.onScroll, {
      capture: true,
      passive: true,
    });
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.onScroll, {
      capture: true,
      passive: true,
    });
  }

  onScroll() {
    if (this.container && !this.props.isLoading) {
      const containerHeight = this.container.clientHeight;
      const scrollTop =
        document.body.scrollTop || document.documentElement.scrollTop;
      const windowHeight =
        window.innerHeight ||
        document.documentElement.clientHeight ||
        document.body.clientHeight;

      if (scrollTop + windowHeight >= containerHeight - SCROLL_THESHOLD) {
        this.loadMore();
      }
    }
  }

  setRef(node) {
    this.container = node;
  }

  loadMore() {
    if (!this.props.isLoading) {
      this.props.loadMore();
    }
  }

  render() {
    return (
      <div ref={this.setRef}>
        {this.props.children}
        {this.props.isLoading && (
          <CenteredContainer>
            <Spinner size="lg" />
          </CenteredContainer>
        )}
      </div>
    );
  }
}

export default InfiniteScoll;
