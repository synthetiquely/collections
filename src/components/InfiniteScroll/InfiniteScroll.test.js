import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import InfiniteScroll from './InfiniteScroll';

describe('InfiniteScroll Component', () => {
  it('should render and match its own snapshot', () => {
    const wrapper = mount(<InfiniteScroll />);
    const tree = renderer.create(<InfiniteScroll />).toJSON();

    expect(wrapper).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });

  it('should evoke loadMore method passed as a props to the component', () => {
    const loadMore = jest.fn();
    const wrapper = mount(<InfiniteScroll loadMore={loadMore} />);

    wrapper.instance().loadMore();

    expect(loadMore).toHaveBeenCalled();
  });
});
