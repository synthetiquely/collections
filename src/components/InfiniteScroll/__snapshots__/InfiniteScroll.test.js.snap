// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll Component should render and match its own snapshot 1`] = `<div />`;
